yeat/tests/test_config.py

```diff
@@ -88,98 +70,174 @@ def test_unsupported_key_in_config_entry():
+    }
+    pattern = rf"Invalid assembly mode '{data['mode']}' for '{label}'"
+    with pytest.raises(AssemblyConfigError, match=pattern):
+        Assembly(label, data, 1)
+
+
+@pytest.mark.parametrize("layer", ["base", "sample", "assembly"])
+def test_check_valid_keys(layer):
+    data = json.load(open(data_file("configs/paired.cfg")))
+    if layer == "base":
+        data["INVALID1"] = ""
+        data["INVALID2"] = ""
+    elif layer == "sample":
+        data["samples"]["Shigella_sonnei_53G"]["INVALID1"] = ""
+        data["samples"]["Shigella_sonnei_53G"]["INVALID2"] = ""
+    elif layer == "assembly":
+        data["assemblies"]["spades-default"]["INVALID1"] = ""
+        data["assemblies"]["spades-default"]["INVALID2"] = ""
+    pattern = r"Found unsupported configuration key\(s\) 'INVALID1,INVALID2'"
+    with pytest.raises(AssemblyConfigError, match=pattern):
+        AssemblyConfig(data, 1)
 
 
 @pytest.mark.parametrize(
-    "readtype,subsamples,sublabels",
-    [
-        (
-            "all",
-            ["sample1", "sample2", "sample3", "sample4"],
-            ["spades-default", "hicanu", "flye_ONT"],
-        ),
-        ("paired", ["sample1", "sample2"], ["spades-default"]),
-        ("pacbio", ["sample3"], ["hicanu"]),
-        ("oxford", ["sample4"], ["flye_ONT"]),
-    ],
+    "mode,label", [("paired", "spades-default"), ("pacbio", "hicanu"), ("oxford", "flye_ONT")]
 )
-def test_to_dict(readtype, subsamples, sublabels):
-    arglist = ["-t", "4", data_file("configs/example.cfg")]
-    args = cli.get_parser().parse_args(arglist)
-    config = AssemblerConfig.from_json(args.config, args.threads)
-    observed = config.to_dict(args, readtype)
-    for sample in subsamples:
-        assert sample in observed["samples"]
-    for label in sublabels:
-        assert label in observed["labels"]
-        assert label in observed["assemblers"]
-        assert observed["assemblers"][label] in label
-    assert len(observed["samples"]) == len(subsamples)
-    assert len(observed["labels"]) == len(sublabels)
-
-
-def test_multiple_readtypes_in_sample():
-    data = json.load(open(data_file("configs/ont.cfg")))
-    data["samples"]["Ecoli_K12_MG1655_R10.3_HAC"]["extra_readtype"] = ["long_read.fastq"]
-    pattern = r"Multiple read types in sample 'Ecoli_K12_MG1655_R10.3_HAC'"
-    with pytest.raises(ValueError, match=pattern):
-        AssemblerConfig(data, 1)
+def test_validate_samples_to_assembly_modes(mode, label):
+    data = json.load(open(data_file("configs/example.cfg")))
+    if mode == "paired":
+        data["assemblies"]["spades-default"]["samples"] = ["sample4"]
+    elif mode == "pacbio":
+        data["assemblies"]["hicanu"]["samples"] = ["sample1"]
+    elif mode == "oxford":
+        data["assemblies"]["flye_ONT"]["samples"] = ["sample1"]
+    pattern = rf"No samples can interact with assembly mode '{mode}' for '{label}'"
+    with pytest.raises(AssemblyConfigError, match=pattern):
+        AssemblyConfig(data, 4)
 
```

Comment:
I'm not 100% sure, but this might be the result of the lack of an `else` clause in the for loop of that case. You could see if that makes a difference. In any case, 100% coverage is a great goal but not always feasible.